File: pet-allergy-scanner/pet-allergy-scanner/Utils/ImageOptimizer.swift
//
//  ImageOptimizer.swift
//  pet-allergy-scanner
//
//  Created by Steven Matos on 10/1/25.
//

import UIKit
import SwiftUI

/// Image optimization utility for compressing and resizing images
/// Ensures images meet size requirements while maintaining quality
struct ImageOptimizer {
    
    /// Maximum file size in bytes (5MB for Supabase free tier)
    static let maxFileSize = 5_242_880 // 5MB
    
    /// Target file size for optimal storage (2MB)
    static let targetFileSize = 2_097_152 // 2MB
    
    /// Maximum image dimension (width or height)
    static let maxDimension: CGFloat = 1024
    
    /// Minimum compression quality (0.0 - 1.0)
    static let minQuality: CGFloat = 0.3
    
    /// Optimize image for upload with smart compression and resizing
    /// - Parameters:
    ///   - image: The source UIImage
    ///   - targetSize: Optional target file size in bytes (defaults to targetFileSize)
    ///   - maxSize: Optional maximum file size in bytes (defaults to maxFileSize)
    /// - Returns: Optimized image data and metadata
    /// - Throws: ImageOptimizationError if optimization fails
    static func optimizeForUpload(
        image: UIImage,
        targetSize: Int = targetFileSize,
        maxSize: Int = maxFileSize
    ) throws -> OptimizedImageResult {
        
        var currentImage = image
        var currentData: Data?
        var currentQuality: CGFloat = 0.85 // Start with high quality
        var compressionAttempts = 0
        let maxAttempts = 8
        
        // Step 1: Resize if image is too large
        if shouldResize(image) {
            currentImage = resizeImage(image, maxDimension: maxDimension)
        }
        
        // Step 2: Initial compression attempt
        currentData = currentImage.jpegData(compressionQuality: currentQuality)
        
        guard var data = currentData else {
            throw ImageOptimizationError.compressionFailed
        }
        
        // Step 3: Progressive compression if needed
        while data.count > targetSize && compressionAttempts < maxAttempts {
            compressionAttempts += 1
            
            // Reduce quality progressively
            currentQuality -= 0.1
            
            // Don't go below minimum quality
            if currentQuality < minQuality {
                currentQuality = minQuality
                
                // If still too large, resize further
                let reductionFactor = sqrt(Double(targetSize) / Double(data.count))
                let newDimension = maxDimension * CGFloat(reductionFactor)
                currentImage = resizeImage(currentImage, maxDimension: newDimension)
            }
            
            guard let compressedData = currentImage.jpegData(compressionQuality: currentQuality) else {
                throw ImageOptimizationError.compressionFailed
            }
            
            data = compressedData
        }
        
        // Step 4: Final size check
        if data.count > maxSize {
            throw ImageOptimizationError.exceedsMaxSize(
                currentSize: data.count,
                maxSize: maxSize
            )
        }
        
        // Calculate compression ratio
        let originalSize = image.jpegData(compressionQuality: 1.0)?.count ?? data.count
        let compressionRatio = Double(data.count) / Double(originalSize)
        
        return OptimizedImageResult(
            data: data,
            finalSize: data.count,
            originalSize: originalSize,
            compressionRatio: compressionRatio,
            finalQuality: currentQuality,
            dimensions: currentImage.size,
            compressionAttempts: compressionAttempts
        )
    }
    
    /// Check if image should be resized based on dimensions
    /// - Parameter image: The image to check
    /// - Returns: True if image exceeds maximum dimensions
    private static func shouldResize(_ image: UIImage) -> Bool {
        return image.size.width > maxDimension || image.size.height > maxDimension
    }
    
    /// Resize image while maintaining aspect ratio
    /// - Parameters:
    ///   - image: The source image
    ///   - maxDimension: Maximum width or height
    /// - Returns: Resized image
    static func resizeImage(_ image: UIImage, maxDimension: CGFloat) -> UIImage {
        let size = image.size
        let aspectRatio = size.width / size.height
        
        var newSize: CGSize
        if size.width > size.height {
            // Landscape
            newSize = CGSize(width: maxDimension, height: maxDimension / aspectRatio)
        } else {
            // Portrait or square
            newSize = CGSize(width: maxDimension * aspectRatio, height: maxDimension)
        }
        
        // Use high-quality rendering
        let renderer = UIGraphicsImageRenderer(size: newSize)
        let resizedImage = renderer.image { context in
            image.draw(in: CGRect(origin: .zero, size: newSize))
        }
        
        return resizedImage
    }
    
    /// Create thumbnail from image
    /// - Parameters:
    ///   - image: Source image
    ///   - size: Thumbnail size (square)
    /// - Returns: Thumbnail image
    static func createThumbnail(from image: UIImage, size: CGFloat = 200) -> UIImage {
        let targetSize = CGSize(width: size, height: size)
        let renderer = UIGraphicsImageRenderer(size: targetSize)
        
        return renderer.image { context in
            let aspectWidth = targetSize.width / image.size.width
            let aspectHeight = targetSize.height / image.size.height
            let aspectRatio = max(aspectWidth, aspectHeight)
            
            let scaledImageSize = CGSize(
                width: image.size.width * aspectRatio,
                height: image.size.height * aspectRatio
            )
            
            let x = (targetSize.width - scaledImageSize.width) / 2.0
            let y = (targetSize.height - scaledImageSize.height) / 2.0
            
            let imageRect = CGRect(
                origin: CGPoint(x: x, y: y),
                size: scaledImageSize
            )
            
            image.draw(in: imageRect)
        }
    }
    
    /// Format file size for display
    /// - Parameter bytes: File size in bytes
    /// - Returns: Formatted string (e.g., "2.5 MB")
    static func formatFileSize(_ bytes: Int) -> String {
        let formatter = ByteCountFormatter()
        formatter.allowedUnits = [.useKB, .useMB]
        formatter.countStyle = .file
        return formatter.string(fromByteCount: Int64(bytes))
    }
}

/// Result of image optimization containing data and metadata
struct OptimizedImageResult {
    /// The optimized image data
    let data: Data
    
    /// Final file size in bytes
    let finalSize: Int
    
    /// Original file size in bytes
    let originalSize: Int
    
    /// Compression ratio (0.0 - 1.0)
    let compressionRatio: Double
    
    /// Final compression quality used
    let finalQuality: CGFloat
    
    /// Final image dimensions
    let dimensions: CGSize
    
    /// Number of compression attempts
    let compressionAttempts: Int
    
    /// Formatted final size string
    var finalSizeFormatted: String {
        ImageOptimizer.formatFileSize(finalSize)
    }
    
    /// Formatted original size string
    var originalSizeFormatted: String {
        ImageOptimizer.formatFileSize(originalSize)
    }
    
    /// Size reduction percentage
    var sizeReduction: Double {
        (1.0 - compressionRatio) * 100.0
    }
    
    /// Human-readable summary
    var summary: String {
        """
        Original: \(originalSizeFormatted)
        Optimized: \(finalSizeFormatted)
        Reduced by: \(String(format: "%.1f%%", sizeReduction))
        Quality: \(String(format: "%.0f%%", finalQuality * 100))
        Size: \(Int(dimensions.width))×\(Int(dimensions.height))
        """
    }
}

/// Image optimization errors
enum ImageOptimizationError: LocalizedError {
    case compressionFailed
    case exceedsMaxSize(currentSize: Int, maxSize: Int)
    case invalidImage
    
    var errorDescription: String? {
        switch self {
        case .compressionFailed:
            return "Failed to compress image"
        case .exceedsMaxSize(let currentSize, let maxSize):
            return "Image size (\(ImageOptimizer.formatFileSize(currentSize))) exceeds maximum allowed (\(ImageOptimizer.formatFileSize(maxSize)))"
        case .invalidImage:
            return "Invalid image format"
        }
    }
}

File: server/app/services/image_optimizer.py
"""
Server-side image optimization service
Handles image compression, resizing, and format conversion
"""

from PIL import Image
from io import BytesIO
from typing import Tuple, Optional
from app.utils.logging_config import get_logger

logger = get_logger(__name__)

class ImageOptimizerService:
    """Service for optimizing images on the server side"""
    
    # Maximum file size in bytes (5MB)
    MAX_FILE_SIZE = 5_242_880
    
    # Target file size for optimal storage (2MB)
    TARGET_FILE_SIZE = 2_097_152
    
    # Maximum image dimension (width or height)
    MAX_DIMENSION = 1024
    
    # Minimum JPEG quality
    MIN_QUALITY = 30
    
    # Starting JPEG quality
    START_QUALITY = 85
    
    @classmethod
    def optimize_image(
        cls,
        image_data: bytes,
        target_size: int = TARGET_FILE_SIZE,
        max_size: int = MAX_FILE_SIZE
    ) -> Tuple[bytes, dict]:
        """
        Optimize image with progressive compression and resizing
        
        Args:
            image_data: Raw image bytes
            target_size: Target file size in bytes
            max_size: Maximum allowed file size in bytes
            
        Returns:
            Tuple of (optimized_image_bytes, metadata_dict)
            
        Raises:
            ValueError: If image cannot be optimized within constraints
        """
        try:
            # Open image
            img = Image.open(BytesIO(image_data))
            original_size = len(image_data)
            original_dimensions = img.size
            
            # Convert to RGB if necessary (handles RGBA, P, L modes)
            if img.mode != 'RGB':
                if img.mode == 'RGBA':
                    # Handle transparency by adding white background
                    background = Image.new('RGB', img.size, (255, 255, 255))
                    background.paste(img, mask=img.split()[3] if len(img.split()) == 4 else None)
                    img = background
                else:
                    img = img.convert('RGB')
            
            # Step 1: Resize if image is too large
            if cls._should_resize(img):
                img = cls._resize_image(img, cls.MAX_DIMENSION)
            
            # Step 2: Progressive compression
            quality = cls.START_QUALITY
            attempts = 0
            max_attempts = 10
            
            while attempts < max_attempts:
                attempts += 1
                
                # Compress image
                output = BytesIO()
                img.save(
                    output,
                    format='JPEG',
                    quality=quality,
                    optimize=True,
                    progressive=True
                )
                compressed_data = output.getvalue()
                current_size = len(compressed_data)
                
                # Check if size is acceptable
                if current_size <= target_size:
                    break
                
                # Reduce quality
                quality -= 10
                
                # If quality too low, resize further
                if quality < cls.MIN_QUALITY:
                    quality = cls.MIN_QUALITY
                    reduction_factor = (target_size / current_size) ** 0.5
                    new_dimension = int(max(img.size) * reduction_factor)
                    img = cls._resize_image(img, new_dimension)
            
            # Final size check
            if current_size > max_size:
                raise ValueError(
                    f"Image size ({current_size:,} bytes) exceeds maximum "
                    f"allowed ({max_size:,} bytes) after {attempts} optimization attempts"
                )
            
            # Calculate metadata
            compression_ratio = current_size / original_size
            size_reduction = (1 - compression_ratio) * 100
            
            metadata = {
                "original_size": original_size,
                "optimized_size": current_size,
                "compression_ratio": round(compression_ratio, 3),
                "size_reduction_percent": round(size_reduction, 1),
                "original_dimensions": {
                    "width": original_dimensions[0],
                    "height": original_dimensions[1]
                },
                "optimized_dimensions": {
                    "width": img.size[0],
                    "height": img.size[1]
                },
                "final_quality": quality,
                "optimization_attempts": attempts
            }
            
            logger.info(
                f"Image optimized: {original_size:,} → {current_size:,} bytes "
                f"({size_reduction:.1f}% reduction)"
            )
            
            return compressed_data, metadata
            
        except Exception as e:
            logger.error(f"Image optimization failed: {str(e)}")
            raise ValueError(f"Image optimization failed: {str(e)}")
    
    @classmethod
    def _should_resize(cls, img: Image.Image) -> bool:
        """Check if image should be resized based on dimensions"""
        return max(img.size) > cls.MAX_DIMENSION
    
    @classmethod
    def _resize_image(cls, img: Image.Image, max_dimension: int) -> Image.Image:
        """
        Resize image while maintaining aspect ratio
        
        Args:
            img: PIL Image object
            max_dimension: Maximum width or height
            
        Returns:
            Resized PIL Image object
        """
        width, height = img.size
        aspect_ratio = width / height
        
        if width > height:
            # Landscape
            new_width = max_dimension
            new_height = int(max_dimension / aspect_ratio)
        else:
            # Portrait or square
            new_width = int(max_dimension * aspect_ratio)
            new_height = max_dimension
        
        # Use high-quality Lanczos resampling
        return img.resize((new_width, new_height), Image.Resampling.LANCZOS)
    
    @classmethod
    def create_thumbnail(
        cls,
        image_data: bytes,
        size: int = 200
    ) -> bytes:
        """
        Create square thumbnail from image
        
        Args:
            image_data: Raw image bytes
            size: Thumbnail size (square)
            
        Returns:
            Thumbnail image bytes
        """
        img = Image.open(BytesIO(image_data))
        
        # Convert to RGB if necessary
        if img.mode != 'RGB':
            if img.mode == 'RGBA':
                background = Image.new('RGB', img.size, (255, 255, 255))
                background.paste(img, mask=img.split()[3] if len(img.split()) == 4 else None)
                img = background
            else:
                img = img.convert('RGB')
        
        # Create thumbnail
        img.thumbnail((size, size), Image.Resampling.LANCZOS)
        
        # Save as JPEG
        output = BytesIO()
        img.save(output, format='JPEG', quality=85, optimize=True)
        return output.getvalue()
    
    @classmethod
    def validate_image(cls, image_data: bytes) -> bool:
        """
        Validate that data is a valid image
        
        Args:
            image_data: Raw image bytes
            
        Returns:
            True if valid image, False otherwise
        """
        try:
            img = Image.open(BytesIO(image_data))
            img.verify()
            return True
        except Exception:
            return False
    
    @classmethod
    def get_image_info(cls, image_data: bytes) -> dict:
        """
        Get information about an image
        
        Args:
            image_data: Raw image bytes
            
        Returns:
            Dictionary with image information
        """
        try:
            img = Image.open(BytesIO(image_data))
            return {
                "format": img.format,
                "mode": img.mode,
                "size": img.size,
                "width": img.size[0],
                "height": img.size[1],
                "file_size": len(image_data)
            }
        except Exception as e:
            raise ValueError(f"Failed to get image info: {str(e)}")

